app/controllers/posts_controller.rb

```diff
@@ -44,16 +43,38 @@ def update
   end
 
   # DELETE /posts/1
-  # DELETE /posts/1.json
   def destroy
     @post.destroy
     redirect_to posts_url, notice: 'Post was successfully destroyed.'
   end
 
+  def favorite
+    unless current_user.has_favorite?(@post)
+      FavoriteHandler.create(user_id: current_user.id, post_id: @post.id )
+    end
+    redirect_to posts_path
+  end
+
+  def unfavorite
+    FavoriteHandler.destroy_favorite(current_user.id, @post.id)
+    redirect_to posts_path
+  end
+
   private
     # Use callbacks to share common setup or constraints between actions.
     def set_post
-      @post = Post.find(params[:id])
+      begin
+        @post = Post.find(params[:id])
+      rescue
+        redirect_to posts_url, alert: 'Post could not be found'
+      end
```

Comment:
ActiveRecord::RecordNotFound exception will render a 404.html page which you can specify in your `public/` directory. You should not be rescuing and rendering alerts like this